[#4214] Keybinds: Open `CompendiumBrowser`, Toggle Sheet Lock

Sayshal opened the pull request:
This PR adds 2 keybinds:

1. **Shift + B**: Opens Compendium Browser in default state.
2. **L**: While in an application with a sheet MODE toggle (Edit/View), toggles that status.

Closes #4214 

Open to discussion on keys.

Sayshal committed:
Add Keybinds

Sayshal committed:
Prevent duplicate CompendiumBrowsers

arbron commented on module/settings.mjs:
```diff
@@ -2,6 +2,7 @@ import CompendiumBrowser from "./applications/compendium-browser.mjs";
 import BastionSettingsConfig from "./applications/settings/bastion-settings.mjs";
 import CalendarSettingsConfig from "./applications/settings/calendar-settings.mjs";
 import CombatSettingsConfig from "./applications/settings/combat-settings.mjs";
+import CompendiumBrowser from "./applications/compendium-browser.mjs";
```
This is already imported at the top.


```suggestion
```

Sayshal replied:
Resolved.

arbron commented on module/settings.mjs:
```diff
@@ -41,6 +42,32 @@ export function registerSystemKeybindings() {
+        const { MODES } = app.constructor;
+        app._mode = app._mode === MODES.EDIT ? MODES.PLAY : MODES.EDIT;
+        app.render();
+      }
```
There is an issue with this method in that it doesn't properly update the tooltip/aria label for the play/edit toggle. The `_onChangeSheetMode` method currently handles this, but it isn't really designed to be called externally. We should instead probably introduce a new `changeMode` public method onto `PrimaryDocumentMixin` and let that handle the label updates:

```javascript
// `applications/api/primary-sheet-mixin.mjs` line 315
    /**
     * Handle changing the sheet mode.
     * @this {PrimarySheet5e}
     * @param {Event} event         Triggering click event.
     * @param {HTMLElement} target  Button that was clicked.
     */
    static #changeMode(event, target) {
      this.changeMode();
    }

    /* -------------------------------------------- */

    /**
     * Handle the user toggling the sheet mode.
     * @param {Event} event         Triggering click event.
     * @param {HTMLElement} target  Button that was clicked.
     * @protected
     */
    async _onChangeSheetMode(event, target=event.currentTarget) {
      foundry.utils.logCompatibilityWarning(
        "The `_onChangeSheetMode` method has been moved to `changeMode`.",
        { since: "DnD5e 6.0", until: "DnD5e 6.2" }
      );
      this.changeMode();
    }

    /* -------------------------------------------- */

    /**
     * Change the sheet mode.
     * @param {PrimarySheetMixin.MODES} [mode]  Mode to set. If not provided, mode will be toggled.
     */
    async changeMode(mode) {
      const { MODES } = this.constructor;
      this._mode = mode ?? (this.isEditMode ? MODES.PLAY : MODES.EDIT);
      const button = this.element?.querySelector('[data-action="changeMode"]');
      if ( button ) {
        const label = _loc(`DND5E.SheetMode${this.isEditMode ? "Edit" : "Play"}`);
        button.checked = this.isEditMode;
        button.dataset.tooltip = label;
        button.setAttribute("aria-label", label);
      }
      await this.submit();
      this.render();
    }
```

That greatly simplifies what we call here:

```suggestion
      if ( app?.rendered && app.changeMode ) app.changeMode();
```

arbron reviewed (changes requested):
I notice these aren't working in detached windows, it seems the keybinding handlers aren't being called at all. That seems like a core issue, but I'll let Kim way in when he reviews.

Sayshal committed:
Address reviewer feedback

Fyorl committed:
Review pass

 - Fixed tooltip label being the wrong way around.
 - Fixed some JSDoc issues.
 - Fixed edit toggle consuming keypress event indiscriminately.
 - Added workaround for core detached window bug.
 - Switched edit mode toggle to Shift+E.

Fyorl reviewed (approved):
Added a workaround for that core detached window bug. I also changed the edit mode toggle from `L` to `Shift+E` for parity with `Shift+B`. Maybe `L` was just one of the few unreserved single letters? Didn't seem related to edit toggle. Maybe edit toggLe? I also considered `F2` since that's used in lots of software for Edit but if you have a keyboard without its own function row that's a bit annoying. Probably doesn't matter too much since you can just rebind it anyway.
